icu4c/source/i18n/number_longnames.cpp

```diff
@@ -119,13 +131,227 @@ static UnicodeString getWithPlural(
+            }
+        }
+    }
+
+  private:
+    bool forCurrentPluralForm(ResourceValue &value, UErrorCode &status) {
+        ResourceTable genderTable = value.getTable(status);
+        if (uprv_strcmp(gender, "") != 0) {
+            if (forGender(genderTable, gender, value, status)) {
+                return true;
+            }
+            if (uprv_strcmp(gender, "neuter") != 0 && forGender(genderTable, "neuter", value, status)) {
+                return true;
+            }
+        }
+        if (forGender(genderTable, "_", value, status)) {
+            return true;
+        }
+        return false;
+    }
+
+    bool forGender(const ResourceTable &genderTable,
+                   const char *genderVal,
+                   ResourceValue &value,
+                   UErrorCode &status) {
+        if (!genderTable.findValue(genderVal, value)) {
+            return false;
+        }
+        ResourceTable caseTable = value.getTable(status);
+        if (uprv_strcmp(caseVariant, "") != 0) {
+            if (forCase(caseTable, caseVariant, value)) {
+                return true;
+            }
+            if (uprv_strcmp(caseVariant, "nominative") != 0 &&
+                forCase(caseTable, "nominative", value)) {
+                return true;
+            }
+        }
+        if (forCase(caseTable, "_", value)) {
+            return true;
+        }
+        return false;
+    }
+
+    bool forCase(const ResourceTable &caseTable,
+                 const char *caseValue,
+                 ResourceValue &value) {
+        if (!caseTable.findValue(caseValue, value)) {
+            return false;
+        }
+        return true;
+    }
+
+    const char *gender;
+    const char *caseVariant;
+    UnicodeString *outArray;
+};
+
+void getInflectedMeasureData(StringPiece subKey,
+                             const Locale &locale,
+                             const UNumberUnitWidth &width,
+                             const char *gender,
+                             const char *caseVariant,
+                             UnicodeString *outArray,
+                             UErrorCode &status) {
+    InflectedPluralSink sink(gender, caseVariant, outArray);
+    LocalUResourceBundlePointer unitsBundle(ures_open(U_ICUDATA_UNIT, locale.getName(), &status));
+    if (U_FAILURE(status)) { return; }
+
+    CharString key;
+    key.append("units", status);
+    if (width == UNUM_UNIT_WIDTH_NARROW) {
+        key.append("Narrow", status);
+    } else if (width == UNUM_UNIT_WIDTH_SHORT) {
+        key.append("Short", status);
+    }
+    key.append("/", status);
+    key.append(subKey, status);
+
+    UErrorCode localStatus = status;
+    ures_getAllItemsWithFallback(unitsBundle.getAlias(), key.data(), sink, status);
+    if (width == UNUM_UNIT_WIDTH_SHORT) {
+        status = localStatus;
+        return;
+    }
+
+    // Fallback to short. (Apparently this somehow happens automatically in
+    // Java? TODO: find out how!)
```

Comment:
The fact that it "works" in Java is an unintentional and not robust: https://unicode-org.atlassian.net/browse/ICU-13353

Reply:
Adjusted the comment, added a TODO(ICU-13353), thanks!